cluster-autoscaler/core/scaledown/actuation/update_latency_tracker_test.go

```diff
@@ -0,0 +1,176 @@
+	"k8s.io/autoscaler/cluster-autoscaler/utils/test"
+)
+
+// mockClock is used to mock time.Now() when testing UpdateLatencyTracker
+// For the n th call to Now() it will return a timestamp after duration[n] to
+// the startTime if n < the length of durations. Otherwise, it will return current time.
+type mockClock struct {
+	startTime time.Time
+	durations []time.Duration
+	index     int
+	mutex     sync.Mutex
+}
+
+// Returns a new NewMockClock object
+func NewMockClock(startTime time.Time, durations []time.Duration) mockClock {
+	return mockClock{
+		startTime: startTime,
+		durations: durations,
+		index:     0,
+	}
+}
+
+// Returns a time after Nth duration from the start time if N < length of durations.
+// Otherwise, returns the current time
+func (m *mockClock) Now() time.Time {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	var timeToSend time.Time
+	if m.index < len(m.durations) {
+		timeToSend = m.startTime.Add(m.durations[m.index])
+	} else {
+		timeToSend = time.Now()
+	}
+	m.index += 1
+	return timeToSend
+}
+
+// Returns the number of times that the Now function was called
+func (m *mockClock) getIndex() int {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	return m.index
+}
+
+// TestCustomNodeLister can be used to mock nodeLister Get call when testing delayed tainting
+type TestCustomNodeLister struct {
+	nodes                    map[string]*apiv1.Node
+	getCallCount             map[string]int
+	nodeTaintAfterNthGetCall map[string]int
+}
+
+// List returns all nodes in test lister.
+func (l *TestCustomNodeLister) List() ([]*apiv1.Node, error) {
+	var nodes []*apiv1.Node
+	for _, node := range l.nodes {
+		nodes = append(nodes, node)
+	}
+	return nodes, nil
+}
+
+// Get returns node from test lister. Add ToBeDeletedTaint to the node
+// during the N th call specified in the nodeTaintAfterNthGetCall
+func (l *TestCustomNodeLister) Get(name string) (*apiv1.Node, error) {
+	for _, node := range l.nodes {
+		l.getCallCount[node.Name] += 1
+		if node.Name == name {
+			if l.getCallCount[node.Name] == l.nodeTaintAfterNthGetCall[node.Name] {
+				toBeDeletedTaint := apiv1.Taint{Key: taints.ToBeDeletedTaint, Effect: apiv1.TaintEffectNoSchedule}
+				node.Spec.Taints = append(node.Spec.Taints, toBeDeletedTaint)
+			}
+			return node, nil
+		}
+	}
+	return nil, fmt.Errorf("Node %s not found", name)
+}
+
+// Return new TestCustomNodeLister object
+func NewTestCustomNodeLister(nodes map[string]*apiv1.Node, nodeTaintAfterNthGetCall map[string]int) *TestCustomNodeLister {
+	getCallCounts := map[string]int{}
+	for name := range nodes {
+		getCallCounts[name] = 0
+	}
+	return &TestCustomNodeLister{
+		nodes:                    nodes,
+		getCallCount:             getCallCounts,
+		nodeTaintAfterNthGetCall: nodeTaintAfterNthGetCall,
+	}
+}
+
+func TestUpdateLatencyCalculation(t *testing.T) {
+
+	testCases := []struct {
+		description string
+		startTime   time.Time
+		// N < 1 result in node will not get tainted before timeout
```

Comment:
nit: Maybe just don't put anything in the map to indicate the node never gets tainted?

Reply:
Done